andi/andi.py

```diff
@@ -1,62 +1,318 @@
+from andi.typeutils import (
+    get_union_args,
+    is_union,
+    get_globalns,
+    get_unannotated_params,
+)
+from andi.errors import CyclicDependencyError, NonProvidableError
 
 
 def inspect(func: Callable) -> Dict[str, List[Optional[Type]]]:
     """
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
         else:
-            res[key] = [tp]
+            res[key] = [] if tp is None else [tp]
     return res
 
 
-def to_provide(
-        arguments_or_func: Union[
-            Callable,
-            Dict[str, List[Optional[Type]]]
-        ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+PlanMapping = MutableMapping[Type, Dict[str, Type]]
+
+
+class Plan(PlanMapping):
+    """
+    The plan resultant of executing the ``plan`` function.
     """
-    Return a dictionary ``{argument_name: type}`` with types which should
-    be provided.
 
-    ``arguments_or_func`` should be either a callable, or
-    a result of andi.inspect call for a callable.
+    def __init__(self):
+        self._plan = OrderedDict()
+
+    def __getitem__(self, item):
+        return self._plan.__getitem__(item)
+
+    def __setitem__(self, key, value):
+        return self._plan.__setitem__(key, value)
+
+    def __delitem__(self, key):
+        return self._plan.__delitem__(key)
+
+    def __iter__(self):
+        return self._plan.__iter__()
+
+    def __len__(self):
+        return self._plan.__len__()
+
+    def __str__(self):
+        return self._plan.__str__()
+
+    def __repr__(self):
+        return self._plan.__repr__()
+
+    @property
+    def dependencies(self) -> PlanMapping:
+        """
+        The plan for build the dependencies of the last task of the plan.
+        Useful when it is known that the last task of the plan could be
+        incomplete, that is, not all dependencies for the last task could be
+        resolved. In such a case is convenient to execute the plan
+        for the dependencies and the have a custom execution for the
+        last task of the plan.
+        """
+        return OrderedDict(list(self.items())[:-1])
+
+    @property
+    def final_arguments(self) -> Dict[str, Type]:
```

Comment:
@kmike not sure about this name but I couldn't come up with something better.

Reply:
final_kwargs?

Reply:
The problem with `kwargs` is that anybody would expect given the name  a dict of `param_name` -> `value`, but here is `param_name` -> `value_builder`. 

`value_builder` is typically the class. So this dict is more some kind of `meta_kwargs`. 

But `final_meta_kwargs` is not a good name in any case. 

`final_args_types`? `final_args_and_types`?

Also the `final` is maybe not a good name. We call it final, but they represent the very first arguments found on inspection. Indeed, they are the argument of the input function or class. They also represent the resolved arguments. 

`resolved_input_args`?

Reply:
Ah, a good catch. In the example in docstring this thing is called kwarg_types, but it is a bit wrong, as a builder can be a function, not a class. final_kwarg_builders?

I don't quite like `args`, because they usually refer to positional arguments (*args, **kwargs)

Reply:
I was having another thought on that. This is the way the user will use this attribute:
```py
plan_steps = plan(fn, is_injectable={A, B})
instances = build(plan_steps.dependencies)
fn(non_annotated='non_annotated',
   **_get_kwargs(instances, plan_steps.final_arguments))
```
The name should be one so that it has sense in this context.
- `final` can mean `definitive` which would fit but also `last` which does not make sense in this context. An alternative could be `settled`, but I don't know if it is a very common word. 
- Using `input` in the name would also be confusing. What is the `input`? `input_class` or `input_func` would be nice, but it could be a class or a function, and `input_class_or_function` is not an option. So discarded. 
- `builders` means that you can use it for building instances, but in this context, you don't invoke them but instead use them as keys for accessing the `instances` dictionary. It is confusing. Even if it is accurate, probably better not be explicit. 
- You don't like `args` because is to close to "`*args, **kwargs`". But using `kwargs` instead wouldn't be also totally accurate. I prefer then `arguments` as it was. 

So then, after thinking I think the original `final_arguments` was not that bad :rofl: . A possible alternative would be `settled_arguments` but I don't know if it is better. What do you think?

Reply:
Re final vs definitive vs last vs settled vs input - my preference is "final" as well, because of the reasons you mentioned.

My problem with `final_arguments` though is that it has the same issue as `final_kwargs` - final_arguments doesn't contain final arguments, it is a `{"name": builder}` dictionary. With `final_kwarg_builders` it is at least clear that the value is a dictionary, but I see your point about `builders`, and agree with it.

Another idea to explore is to get rid of final_arguments and alike helpers, and provide a different helper, something like this:

```py
class Plan(List[Steps]):
    def final_kwargs(self, instances):
        return _get_kwargs(instances, self.steps[-1][1])
```

Reply:
@kmike interesting idea. I have explored it a little bit and come up with a proposal. I created a separated PR: https://github.com/scrapinghub/andi/pull/12 so that you can see clearly the proposed changes, but if you think is a good way to go, better to merge here as soon as possible and discuss possible required changes in this PR. 

Mainly the idea is to introduce the concept of kwargs specification.